runners/direct-java/src/main/java/org/apache/beam/runners/direct/TransformExecutor.java

```diff
@@ -104,19 +104,23 @@ public void run() {
         return;
       }
 
-      processElements(evaluator, metricsContainer, enforcements);
+      processElements(evaluator, enforcements);
 
-      finishBundle(evaluator, metricsContainer, enforcements);
+      finishBundle(evaluator, enforcements);
+
+      synchronized (context) {
+        context.getCommittedMetrics().update(stepName, metricsContainer);
+      }
     } catch (Exception e) {
       onComplete.handleException(inputBundle, e);
       if (e instanceof RuntimeException) {
         throw (RuntimeException) e;
       }
       throw new RuntimeException(e);
     } finally {
-      // Report the physical metrics from the end of this step.
-      context.getMetrics().commitPhysical(inputBundle, metricsContainer.getCumulative());
-
+      synchronized (context) {
```

Comment:
Ah, I forgot this was in the DirectRunner. That sounds reasonable, although @tgroh may have an idea if there is somewhere better to merge containers.